pkg/query/plan/pipeline.go

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2023, Dana Burkart <dana.burkart@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+package plan
+
+import (
+	"fmt"
+	"github.com/dburkart/fossil/pkg/database"
+	"github.com/dburkart/fossil/pkg/query/ast"
+	"github.com/dburkart/fossil/pkg/query/types"
+	"sync"
+)
+
+type DataPipeline interface {
+	Execute(entries database.Entries) database.Entries
+}
+
+type Pipeline struct {
+	stages []Stage
+}
+
+func MakePipelineFromNode(node *ast.DataPipelineNode) Pipeline {
+	var p Pipeline
+
+	for _, stage := range node.Stages {
+		stage, ok := stage.(*ast.DataFunctionNode)
+		if !ok {
+			panic("Unexpected node found in data pipeline")
+		}
+
+		switch stage.Name.Lexeme {
+		case "filter":
+			p.Add(MakeFilterStage(stage))
+		case "map":
+			p.Add(MakeMapStage(stage))
+		case "reduce":
+			p.Add(MakeReduceStage(stage))
+		default:
+			panic(fmt.Sprintf("Unsupported stage type: %s", stage.Name.Lexeme))
+		}
+	}
+	p.Finalize()
+
+	return p
+}
+
+func (p *Pipeline) Add(s Stage) {
+	if len(p.stages) > 0 {
+		p.stages[len(p.stages)-1].Chain(s)
+	}
+	p.stages = append(p.stages, s)
+}
+
+func (p *Pipeline) Finalize() {
+	collect := MakeCollectStage()
+	if len(p.stages) > 0 {
+		p.stages[len(p.stages)-1].Chain(collect)
+	}
+	p.stages = append(p.stages, collect)
+}
+
+func (p *Pipeline) Execute(entries database.Entries) database.Entries {
+	var results database.Entries
+	var wg sync.WaitGroup
+
+	// Start our pipeline stages
+	for _, stage := range p.stages {
+		go stage.Execute()
+	}
+
+	first := p.stages[0]
+	last := p.stages[len(p.stages)-1].(*CollectStage)
+
+	// Start our reader for output
+	wg.Add(1)
+	go func() {
+		for result := range last.Output {
+			results = append(results, result.Entry())
+		}
+		wg.Done()
+	}()
+
+	// Pass in everything to the first stage
+	for _, entry := range entries {
+		first.Add([]WrappedEntry{Wrap(entry)})
+	}
+	first.Finish()
+
+	wg.Wait()
```

Comment:
Do you think we could encounter timeouts or blocking on the database due to a bad pipeline?

Reply:
I don't know 🤷 

I think it could happen if we write a bad stage (i.e. the go code for a stage is bad), but I'm not sure how it could happen from user input. Each stage should close the next when it's done processing (even if it didn't forward anything).

But I haven't tested it on large amounts of data, so I'm not super confident.